### PR TITLE
refactor: Clarify error when right_path exists

### LINKED ISSUE #1552: Python sqlalchemy_oso translate_compare difficult to understand error due to assertion
When writing queries using `sqlalchemy_oso`, if the `assert not right_path` is hit, a hard to understand error is triggered:

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/fastapi/applications.py", line 261, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 23, in __call__
    await responder(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 42, in __call__
    await self.app(scope, receive, self.send_with_gzip)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/cors.py", line 92, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/middleware/cors.py", line 147, in simple_response
    await self.app(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 656, in __call__
    await route.handle(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 259, in handle
    await self.app(scope, receive, send)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 61, in app
    response = await func(request)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/anyio/to_thread.py", line 28, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(func, *args, cancellable=cancellable,
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 818, in run_sync_in_worker_thread
    return await future
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 754, in run
    result = context.run(func, *args)
  File "/Users/username-here/git/docker/projectname-here/api/src/projectname-here/routers/cloud/security_groups/rules.py", line 49, in invalid_ip_security_group_rules_count
    return get_invalid_ip_security_group_rules_count(
  File "/Users/username-here/git/docker/projectname-here/api/src/projectname-here/crud/security_group_rules.py", line 23, in get_invalid_ip_security_group_rules_count
    return cast(int, query.order_by(SecurityGroupRule.id).scalar())
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2888, in scalar
    ret = self.one()
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2865, in one
    return self._iter().one()
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2903, in _iter
    result = self.session.execute(
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1673, in execute
    result = fn(orm_exec_state)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/session.py", line 311, in do_orm_execute
    filter = authorize_model(oso, user, action, session, entity)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/auth.py", line 95, in authorize_model
    filter = partial_to_filter(
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/partial.py", line 122, in partial_to_filter
    return translate_expr(expression, session, model, get_model)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/partial.py", line 137, in translate_expr
    return translate_and(expression, session, model, get_model)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/partial.py", line 152, in translate_and
    translated = translate_expr(expression, session, model, get_model)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/partial.py", line 131, in translate_expr
    return translate_compare(expression, session, model, get_model)
  File "/Users/username-here/Library/Caches/pypoetry/virtualenvs/projectname-here-259ubUSQ-py3.10/lib/python3.10/site-packages/sqlalchemy_oso/partial.py", line 221, in translate_compare
    assert not right_path
AssertionError
```

When debugging, I wrapped the `assert` in a `try/except`

```python
        try:
            assert not right_path
        except AssertionError as e:
            print(f"{right=}")
            print(f"{right_path=}")
            raise e from e
```

and see that right and right path are:

```
right=Expression(Dot, [Expression(And, [Expression(Isa, [Variable('_this'), Pattern(SecurityGroup, {})]), Expression(Unify, [Expression(Dot, [Expression(And, [Expression(Isa, [Variable('_this'), Pattern(Platform, {})]), Expression(Unify, ['core', Expression(Dot, [Variable('_this'), 'name'])])]), 'name']), Expression(Dot, [Variable('_this'), 'platform_name'])])]), 'inventory_id'])
right_path=('inventory_id',)
```

### PR DESCRIPTION
**Pull request by kkirsche**
Fixes #1552

This commit catches the assertion error when verifying the `right_path`
does not exist when a path has been identified on the left side of the
policy. This refactor was made to better clarify the cause of the error
for developers who may not be familiar with the technical limitations of
the current SQLAlchemy integration.

In my case, I had assumed that the use of foreign keys and paths on both
sides would be more performant, and thus the best choice for a policy,
when writing the authorization rules. The resulting error didn't include
any context about what was wrong with the policy I had written. This
commit aims to address this by providing a best effort guess to provide
a starting point for developers to debug the policy issue themselves.

PR checklist:

- [ ] Added changelog entry.

### CODE REVIEW

**Comment by kkirsche** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by gw000** [comment]
Hi @kkirsche , how about
```
if not right_path:
    raise PolarException(...)
```
instead of catching the assertion error?

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py
Comment (gw000) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py:224]: I think you mean `PolarRuntimeError`?

**Review by gw000 — CHANGES_REQUESTED** [review]
Fix the error type, replace `try/assert/except/raise` with `if/raise` and this is good to go 😃

**Comment by kkirsche** [comment]
> Hi @kkirsche , how about
> 
> ```
> if not right_path:
>     raise PolarException(...)
> ```
> 
> instead of catching the assertion error?

Great point, I accidentally missed that.

That and the specific error being raised should be fixed in eb7c26e. Sorry again about the mix up there, but thank you for catching it 👍

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py
Comment (gw000) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py:225]: `e` is undefined

**Review by gw000 — CHANGES_REQUESTED** [review]
One more change sorry!

**Comment by gw000** [comment]
There's still a test that's failing. Are you able to fix it?

**Comment by kkirsche** [comment]
I'll take a look once the tests run after the latest merge of `main` into this branch, @gw000, and I will see if this is something I can clean up or if it's something that the team may be able to assist me in fixing. I will need to do it iteratively due to the errors with the toolchain I was experiencing. Once I see the error in the Github Action, I'll have a better understanding of what will be involved in this change.

Sorry for the delay in responding. A few time-sensitive items came up that I had to work through. I appreciate your patience.

**Comment by kkirsche** [comment]
I'll see if I can fix it, but I can't run tests at all locally so this will come down to how quickly does trying to work with the toolchain become frustrating and I give up on it.

If someone with a deeper understanding of your toolchain is able to assist, that would be hugely helpful for this (and I can grant access to the branch / repo for that help) or I'll probably end up closing this PR if I can't figure it out as the toolchain is sufficiently complex that I can't really do local development for small contributions like this

**Comment by kkirsche** [comment]
```
In [1]: assert not ()

In [2]: if not ():
   ...:     print("HERE")
   ...:
HERE
```

Ah, the criteria for assert and if are different, so the right path of `()` in the if statement is triggering the new error while the old assert ignored this

**Comment by kkirsche** [comment]
The force pushes above were due to me realizing a better approach. Originally, I had updated this to use:

```python
if len(right_path) > 0:
```

because `dot_path` returns a `Tuple` type, this would work. Once I stepped back for a second though, I realized all I had done was swap the conditional's negation.

The force push updated this to:

```python
if right_path
```

The final one fixed the commit message so that it properly reflected the change that was included. Theoretically, if I understood this properly, the tests should pass

**Comment by kkirsche** [comment]
Branch updated based on main — PR is pending re-review.

**Comment by kkirsche** [comment]
Branch updated based on main — PR is pending re-review by a member of the Oso team.

**Comment by kkirsche** [comment]
Course, f I hit this, other people may also run into this, so the more we can do to help them, the better 👍

**Pull request merged**
